<div style="border:solid green 2px; padding: 20px">
    
<b>Александр, привет!</b> Мы рады тебя видеть на территории код-ревьюеров. Ты проделал большую работу над проектом, но давай познакомимся и сделаем его еще лучше! У нас тут своя атмосфера и несколько правил:


1. Меня зовут Александр Матвеевский. Я работаю код-ревьюером, моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать аналитиком данных.
2. Общаемся на ты.
3. Если хочешь написать, спросить - не нужно стесняться. Только выбери свой цвет для комментария.  
4. Это учебный проект, тут можно не бояться сделать ошибку.  
5. У тебя неограниченное количество попыток для сдачи проекта.  
6. Let's Go!

---

Я буду красить комментарии цветом, пожалуйста, не удаляй их:

<div class="alert alert-block alert-danger">✍
    

__Комментарий от ревьюера №1__

Такой комментарий нужно исправить обязательно, он критически влияет на удачное выполнение проекта.
</div>
    
---

<div class="alert alert-block alert-warning">📝
    

__Комментарий от ревьюера №1__


Такой комментарий является рекомендацией или советом. Можешь использовать их на своё усмотрение.
</div>

---

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

Такой комментарий  говорит о том, что было сделано что-то качественное и правильное =)
</div>
    
---
    
Предлагаю работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комментарии — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:   
    
<div class="alert alert-info"> <b>Комментарии студента:</b> Например, вот так.</div>
    
Всё это поможет выполнить повторную проверку твоего проекта оперативнее. Если будут какие-нибудь вопросы по моим комментариям, пиши, будем разбираться вместе :)    
    
---

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
data_pivot = data.pivot_table(index=['children'], values='debt', aggfunc = ['count', 'sum', 'mean','median'])

print(data_pivot.head(10)) # Ваш код будет здесь. Вы можете создавать новые ячейки.

          count   sum      mean median
           debt  debt      debt   debt
children                              
0         14091  1063  0.075438      0
1          4808   444  0.092346      0
2          2052   194  0.094542      0
3           330    27  0.081818      0
4            41     4  0.097561      0
5             9     0  0.000000      0


**Вывод:**Зависимость между количеством детей и возвратом кредита в срок не ярко выражена. В первую очередь необходимо обратить внимание на отсутствие задолженности у семей с 5 детьми (возможно, это ошибка).Возможно, необходимо получить больше информации из других источников о задолженности по кредитам у многодетных семей. Чуть меньше чем остальные категории задолженность имели семьи без детей (7,5%), вследствие небольшой финансовой нагрузки. Семьи с 1,2 и 4 детьми имеют самые худшие показатели по возврату задолженности (9,2%, 9,5% и 9,8% соответственно).

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

Здорово, что посчитал колличество заемщиков в каждой выборке, и учел неравнозначность выборок. 
    

</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
data_pivot = data.pivot_table(index=['family_status'], values='debt', aggfunc = ['count', 'sum', 'mean','median'])

print(data_pivot.head(10)) # Ваш код будет здесь. Вы можете создавать новые ячейки.

                       count  sum      mean median
                        debt debt      debt   debt
family_status                                     
Не женат / не замужем   2796  273  0.097639      0
в разводе               1189   84  0.070648      0
вдовец / вдова           951   63  0.066246      0
гражданский брак        4134  385  0.093130      0
женат / замужем        12261  927  0.075606      0


**Вывод:**Самую лучшую платежную дисциплину имеют люди, состоящие в категории "вдовец / вдова" (6,6%). Возможно, для полноты картины, необходимо произвести сравнение о наличии у людей данной категории детей (количества детей). Категории "женат / замужем" и  "Не женат / не замужем" имеют показатели чуть хуже (7,6% и 7,1% соответственно). РАЗНИЦА МЕЖДУ ЭТИМИ КАТЕГОРИЯМИ СОСТАВЛЯЕТ не БОЛЕЕ 1%. Самую худшую платежную дисциплину имеют люди, состоящие в гражданском браке и не женатые люди (9,3% и 9,8% соответственно). Вероятной причиной этого является то, что люди в этих категория  недостаточно обеспечены ввиду незрелого возраста (гипотеза нуждается в проверке!).

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

Верно. по расчетам можно сказать, что заемщики которые состоят или состояли в браке лучше оплачивают кредиты.
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
data_pivot = data.pivot_table(index=['total_income_category'], values='debt', aggfunc = ['count', 'sum', 'mean','median'])

print(data_pivot.head(10))## Ваш код будет здесь. Вы можете создавать новые ячейки.

                       count   sum      mean median
                        debt  debt      debt   debt
total_income_category                              
A                         25     2  0.080000      0
B                       5014   354  0.070602      0
C                      15921  1353  0.084982      0
D                        349    21  0.060172      0
E                         22     2  0.090909      0


**Вывод:** 
Категории уровня зарплат
0–30000 — 'E';
30001–50000 — 'D';
50001–200000 — 'C';
200001–1000000 — 'B';
1000001 и выше — 'A'.

Как и ожидалось самый высокий риск невзврата задолженности по кредитам у людей с доходом от 0 до 30 000 - 9,09%. Самый высокий процент невозвата у категории "А" (1000001 и выше) (8%) и категории "С" (50001–200000) (8,4%). Самый низкий показаель  и людей, находящихся в категории "D" (30 001 - 50 000) - 6%.
Считанныю, что полученные данные не могут учитыватся без указания процентного соотношения к дохода к задолженности. Категории "B" и "C" являются основными потребителями(заемщиками).

<div class="alert alert-block alert-warning">📝
    

__Комментарий от ревьюера №1__

Какую можно выявить зависимость, если сравнить две самые многочисленные категории?
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
data['purpose'].unique() # У меня вопрос - как посчитать количество уникальных "целей получения кредита"? 
#Просто фунция uniqe() не подходит?

data_pivot = data.pivot_table(index=['purpose_category'], values='debt', aggfunc = ['count', 'sum', 'mean','median'])

print(data_pivot)## Ваш код будет здесь. Вы можете создавать новые ячейки.

                          count  sum      mean median
                           debt debt      debt   debt
purpose_category                                     
операции с автомобилем     4279  400  0.093480      0
операции с недвижимостью  10751  780  0.072551      0
получение образования      3988  369  0.092528      0
проведение свадьбы         2313  183  0.079118      0


**Вывод:** Самый меньший процент невозврата наблюдается в кредитах, предоставленных на операции с недвижимостью (7,2%). Это может быть связано с долгосрочность данных кредитов, и в том числе, из-за того, что данные кредиты берутся под залог приобретаемой недвижимости (данную недвижимость можно реализовать для оплаты по кредиту). (Интересно было посмотреть источники погашения кредитов(личные средства или средства от продажи недвижимости).
Далее идет кредиты на проведения свадьбы (7.9%).Вероятно по причине небольших сумм и сроков по этим кредитам.
Самые плохие показатели по возврату задолженности приходятся на операции с автомобилями (9,3%) и получения образования (9,3%).
Причиной может быть то, что покупка автомобиля — это не инвестиция, как в случае с недвижимостью, а всегда только затраты. "Кредитный автомобиль — это вдвойне затратно" (расходы на страховку, ГСВ и ремонт). Также с проблемой по выплату кредита могут столкнутся автовладельцы дорогих автомобилей, которые, вполне возможно, взяли транспортное средство, которое им не по карману.
Кредиты на получения образования также имеют высокий уровень невозврата, вероятно, в связи с тем, что данные кредиты зачастую получают студенты и пр. имеющие невысокий уровень дохода, либо его отсутствие.



<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

Да, по расчетам можно сделать вывод, что самыми проблемными выглядят автокредиты и кредиты на образование.
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:*   По всей видимости, причина в наличии пропусков в столбцах является то, что заявитель намеренно или случайно не сумму получаемых доходов и количества дней занятости (не все знают эти данные). Также, причиной пропусков может являтся технический фактор (например, ошибка при выгрузке).

<div class="alert alert-block alert-warning">📝
    

__Комментарий от ревьюера №1__

Так же, стоит учитывать технический фактор (например, ошибка при выгрузке) также может стать причиной пропусков в данных.
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:*  Среднее значение некорректно характеризует данные, когда некоторые значения сильно выделяются среди большинства. 

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

Верно, медиана менее чувствительна к выбросам в данных.
</div>

### Шаг 4: общий вывод.

На основании полученных данных, невозможно определить влияние на возврат задолженности. Необходимо более детально провести анализ категорий заемщиков, суммы кредита и источников погашения (отмечно в тексте). 

<div class="alert alert-block alert-warning">📝
    

__Комментарий от ревьюера №1__

Шикарный вывод и рекомендации для бизнеса. Кстати, хорошей практикой является написание в общем выводе всех твоих действий (кратко) по проекту. Ведь вывод раскрывает для заказчика всю твою проделанную работу и за что он платит деньги, даже если ему не интересен весь процесс исследования, то в отчёт он заглянет обязательно. Ты молодец!
</div>

<div style="border:solid blue 3px; padding: 20px">
<div class="alert alert-block alert-success">✔️
Комментарий от ревьюера №1 </b> 



У тебя получилась очень сильная и хорошая работа. Здорово, что не забываешь про комментарии, твой проект интересно проверять. 

---

Нужно поправить:

1) Ничего, отличная работа!)

----


Если у тебя будут какие-то вопросы по моим комментариям - обязательно пиши! Буду ждать работу на повторное ревью :)</div>

<div style="border:solid blue 3px; padding: 20px">
<div class="alert alert-block alert-success">✔️
    

__Коментарий от ревьюера №2__


В остальном всё чудно😊. Твой проект так и просится на github =)   
    
Поздравляю с успешным завершением проекта 😊👍
И желаю успехов в новых работах 😊